### PR TITLE
Feature/xaptc 36 

### PR DESCRIPTION
**Pull request by frankquinn**
PUT Services for Configurations

### CODE REVIEW

**Comment by frankquinn** [comment]
This story has the GET request (XAPTC -35) story pulled into it.

File: uiServices/src/services.go [hunk 1]
```diff
@@ -96,5 +94,141 @@ func validateJsonWithSchema(config []byte) bool {
...
+
+    configPathDir := req.Header.Get("configPathDir")
+    configName := chi.URLParam(req, "configName")
+
+    if !strings.HasSuffix(configPathDir, "/"){
+        configPathDir = configPathDir + "/"
+    }
+
+    file, err := os.Open(fmt.Sprintf("%s%s.xml", configPathDir, configName))
+    if len(configPathDir) <= 1{
```
Comment (jenny-flynn) [inline uiServices/src/services.go]: move the configPathDir length check to before you try to open the file. Note: I flagged this in https://github.com/xtracdev/automated-perf-test/pull/40 , so you can make the change there and pull it in to this branch.

File: uiServices/src/services.go [hunk 2]
```diff
@@ -96,5 +94,141 @@ func validateJsonWithSchema(config []byte) bool {
...
+    }
+
+    rw.WriteHeader(http.StatusOK)
+    rw.Write(configJson)
+    logrus.Println(string(configJson))
 
 }
+
+func putConfigs(rw http.ResponseWriter, req *http.Request) {
+    path := req.Header.Get("configPathDir")
```
Comment (jenny-flynn) [inline uiServices/src/services.go]: enhancement:
create a function "getConfigPathHeader" which returns the header string value and an error if not found. Then you can move a lot of the logic(below) into a separate function that is shared between getConfigs and PutConfigs.

    path := req.Header.Get("configPathDir")
      if !strings.HasSuffix(path, "/") {
             path = path + "/"
         }
 
     if len(path) <= 1 {
         logrus.Error("No Path Entered")
         rw.WriteHeader(http.StatusBadRequest)
         return
     }

File: uiServices/src/services.go [hunk 3]
```diff
@@ -96,5 +94,141 @@ func validateJsonWithSchema(config []byte) bool {
...
+
+    }
+
+    if configName != config.APIName {
+        logrus.Error("Api Name must match File Name")
+        rw.WriteHeader(http.StatusBadRequest)
+        return
+    }
+    //cannot have update with invalid or empty values
+    if config.APIName == "" ||
```
Comment (jenny-flynn) [inline uiServices/src/services.go]: is this reachable if we validate against the json schema? I would suggest we re-evaluate this section.

**Comment by frankquinn** [comment]
resolved conflicts

**Comment by jenny-flynn** [comment]
Can you run `go vet` on this project and address the findings? I'm seeing:

uiServices/src/services_test.go:91: no formatting directive in Errorf call
uiServices/src/services_test.go:114: no formatting directive in Errorf call
uiServices/src/services_test.go:135: no formatting directive in Errorf call
uiServices/src/services_test.go:156: no formatting directive in Errorf call

**Comment by frankquinn** [comment]
Not 100% why that happens.

The code causing it is things like :
 if w.Code != http.StatusCreated {
		t.Errorf("TestValidJsonPost. Expected:", http.StatusCreated, " Got:", w.Code, "  Error. Did not succesfully post")
	}

I'll just change them all to assert.Equal() and that will solve it

**Comment by frankquinn** [comment]
Changed t.Errorf() to t.Error().   go vet doesn't return any error now.

File: uiServices/src/services.go [hunk 4]
```diff
@@ -15,11 +15,38 @@ import (
...
+        rw.WriteHeader(http.StatusNotFound)
+        return
+    }
+
+    if len(header) <= 1 {
+        logrus.Error("No Header Path Found")
+        rw.WriteHeader(http.StatusBadRequest)
+        return
+    }
+    return
```
Comment (jenny-flynn) [inline uiServices/src/services.go]: line not necessary.

File: uiServices/src/services.go [hunk 5]
```diff
@@ -98,53 +125,89 @@ func validateJsonWithSchema(config []byte) bool {
...
+        logrus.Println(string(configJson))
+
+
+}
+
+func putConfigs(rw http.ResponseWriter, req *http.Request) {
+    path := getConfigHeader(req)
+    configName := chi.URLParam(req, "configName")
+
+    validateFileNameAndHeader(rw, req,path,configName)
```
Comment (jenny-flynn) [inline uiServices/src/services.go]: don't know if this logic makes sense. if file name or header are invalid, we want to immediately return. I think at the moment, the flow will continue on through the code. perhaps in this `validateFileNameAndHeader` set the response status code as you're doing, then return a boolean and if false, just return.

File: uiServices/src/services.go [hunk 6]
```diff
@@ -98,53 +125,89 @@ func validateJsonWithSchema(config []byte) bool {
...
         return
     }
 
-    file, err := os.Open(fmt.Sprintf("%s%s.xml", configPathDir, configName))
+    config := perfTestUtils.Config{}
+    err := json.Unmarshal(buf.Bytes(), &config)
     if err != nil {
-        logrus.Error("Configuration Name Not Found: "+configPathDir + configName)
-        rw.WriteHeader(http.StatusNotFound)
+        rw.WriteHeader(http.StatusBadRequest)
```
Comment (jenny-flynn) [inline uiServices/src/services.go]: log that you were unable to unmarshall the json

File: uiServices/src/services.go [hunk 7]
```diff
@@ -98,53 +125,89 @@ func validateJsonWithSchema(config []byte) bool {
...
+        rw.WriteHeader(http.StatusConflict)
         return
+
     }
 
-    err = xml.Unmarshal(byteValue, &config)
-    if err != nil{
-        rw.WriteHeader(http.StatusInternalServerError)
-        logrus.Error("Cannot Unmarshall")
+    if configName != config.APIName {
```
Comment (jenny-flynn) [inline uiServices/src/services.go]: these do not need to match.

**Comment by frankquinn** [comment]
I fixed the PUT issue and the POST issue and have tests for them

File: uiServices/src/services.go [hunk 8]
```diff
@@ -98,53 +132,88 @@ func validateJsonWithSchema(config []byte) bool {
...
+        }
+
+        rw.WriteHeader(http.StatusOK)
+        rw.Write(configJson)
+        logrus.Println(string(configJson))
+
+
+}
+
+func putConfigs(rw http.ResponseWriter, req *http.Request) {
```
Comment (jenny-flynn) [inline uiServices/src/services.go]: run `go fmt ./..`. at the automated-perf-test level

Comment (frankquinn) [reply uiServices/src/services.go]: Not getting any errors from go fmt

image

**Pull request merged**